cmd/list.go

```diff
@@ -0,0 +1,180 @@
+package cmd
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/Benchkram/errz"
+	"github.com/spf13/cobra"
+
+	"github.com/puppetlabs/wash/api/client"
+	apitypes "github.com/puppetlabs/wash/api/types"
+	cmdutil "github.com/puppetlabs/wash/cmd/util"
+	"github.com/puppetlabs/wash/config"
+	"github.com/puppetlabs/wash/plugin"
+)
+
+func listCommand() *cobra.Command {
+	listCmd := &cobra.Command{
+		Use:     "list [file]",
+		Aliases: []string{"ls"},
+		Short:   "Lists the resources at the indicated path.",
+		Args:    cobra.MaximumNArgs(1),
+	}
+
+	listCmd.RunE = toRunE(listMain)
+
+	return listCmd
+}
+
+func headers(timey string) []cmdutil.ColumnHeader {
+	return []cmdutil.ColumnHeader{
+		{ShortName: "name", FullName: "NAME"},
+		{ShortName: "time", FullName: timey},
+		{ShortName: "verbs", FullName: "ACTIONS"},
+	}
+}
+
+func format(t time.Time) string {
+	return t.Format(time.RFC822)
+}
+
+func formatListEntries(apiPath string, ls []apitypes.ListEntry) string {
+	table := make([][]string, len(ls))
+	for i, entry := range ls {
+		var ctimeStr string
+		if entry.Attributes.HasCtime() {
+			ctimeStr = format(entry.Attributes.Ctime())
+		} else {
+			ctimeStr = "<unknown>"
+		}
+
+		verbs := strings.Join(entry.Actions, ", ")
+
+		name := entry.CName
+		if len(ls) > 1 && entry.Path == apiPath {
+			// Represent the pwd as "."
+			name = "."
+		}
+
+		if isListable(entry) {
+			name += "/"
+		}
+
+		table[i] = []string{name, ctimeStr, verbs}
+	}
+	return cmdutil.FormatTable(headers("CREATED"), table)
+}
+
+func findEntry(entries []apitypes.ListEntry, name string) apitypes.ListEntry {
+	for _, entry := range entries {
+		if entry.CName == name {
+			return entry
+		}
+	}
+	return apitypes.ListEntry{}
+}
+
+func isListable(entry apitypes.ListEntry) bool {
+	for _, action := range entry.Actions {
+		if action == plugin.ListAction.Name {
+			return true
+		}
+	}
+	return false
+}
+
+func listResource(apiPath string) error {
+	conn := client.ForUNIXSocket(config.Socket)
+
+	var entries []apitypes.ListEntry
+	if apiPath == "/" {
+		// The root, definitely listable
+		ls, err := conn.List(apiPath)
+		if err != nil {
+			return err
+		}
+		entries = ls
+	} else {
+		// List the parent to see whether it's a single entry or a listable resource
+		parent, base := filepath.Split(apiPath)
```

Comment:
Think this might be a good time to implement that `fs/info/<entry>` endpoint, because the casing here's a bit weird.

Reply:
I'm starting to think that'd make the `List` endpoint cleaner because we can avoid returning `.` in the results. That way, people can see that `List` is strictly meant to invoke the `List` action.

Reply:
That's fair, I think it would be useful.